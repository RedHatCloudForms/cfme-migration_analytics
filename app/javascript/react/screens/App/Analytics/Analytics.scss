#migration-analytics {
  margin-top: 15px;

  &.row.cards-pf {
    margin-top: 0;
    overflow: auto;
    overflow-x: hidden;
    padding-bottom: 50px;
    height: 100%;
  }

  &.cards-pf .blank-slate-pf {
    border: none;

    .blank-slate-pf-icon {
      opacity: 0.4;
    }
  }

  .large-spinner {
    padding: 50px;
    text-align: center;

    .spinner,
    h3 {
      vertical-align: middle;
    }

    h3 {
      display: inline-block;
      margin: 0;
      padding-left: 10px;
    }
  }

  .reports-summary {
    .environment-summary {
      margin-left: 16px;
    }

    table {
      padding-top: 10px;

      td {
        padding: 5px 32px;
      }
    }

    .list-view-pf-expand {
      margin-top: 25px;
    }
  }

  .footer-buttons button {
    margin-right: 10px;
  }

  .detailed-data-selection {
    margin-top: 15px;

    li {
      list-style: disc outside none;
    }
  }

  .icon-with-content {
    display: flex;

    &.data-collection-status {
      justify-content: center;
      padding: 50px;
    }

    &.smart-state-info {
      padding-bottom: 10px;
    }

    > .pficon.p-aligned {
      margin-top: 3px;
    }

    > div,
    > p {
      display: inline-block;
      margin: 0 8px;
    }

    .checkmark {
      font-size: 48px;
      margin-top: 10px;
      color: $color-pf-green;
    }

    h3.beside-spinner {
      margin: 8px 0 30px 0;
    }

    .buttons {
      margin-top: 20px;

      button {
        display: block;
        margin: 15px 0;
      }
    }
  }

  .payload-path {
    display: inline-block;
    padding-left: 20px;
  }
}

#migration-analytics-manifest-version {
  position: fixed;
  bottom: 0;
  right: 0;
  margin: 0;
  padding: 15px;
  background-color: #ffffff;

  &.on-empty-state {
    background-color: $color-pf-black-150;
  }

  > button.btn-link {
    margin-bottom: 2px;
  }

  > .spinner-xs {
    position: relative;
    top: 2px;
    margin-left: 5px;
  }
}

.manifest-update-modal .col-xs-6 {
  padding-left: 5px;
  padding-right: 5px;
}
