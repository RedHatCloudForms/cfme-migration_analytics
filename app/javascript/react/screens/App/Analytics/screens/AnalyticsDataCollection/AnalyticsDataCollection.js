import React from 'react';
import PropTypes from 'prop-types';
import { Spinner, Button } from 'patternfly-react';

const AnalyticsDataCollection = ({ onCancelClick }) => (
  <React.Fragment>
    <div className="large-spinner">
      <Spinner loading size="lg" inline />
      <h3>{__('Collecting inventory data')}</h3>
    </div>
    <Button onClick={onCancelClick}>{__('Cancel')}</Button>
  </React.Fragment>
);

AnalyticsDataCollection.propTypes = {
  onCancelClick: PropTypes.func.isRequired
};

export default AnalyticsDataCollection;
