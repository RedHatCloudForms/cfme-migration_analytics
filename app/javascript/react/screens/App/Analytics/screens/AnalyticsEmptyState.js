import React from 'react';
import PropTypes from 'prop-types';
import { EmptyState, Button } from 'patternfly-react';
import ProcessImprovementSvg from '../components/process-improvement.svg';

const AnalyticsEmptyState = ({ onGetStartedClick }) => (
  <EmptyState className="full-page-empty">
    <div className="blank-slate-pf-icon">
      <ProcessImprovementSvg height="80px" />
    </div>
    <EmptyState.Title>{__('Examine your virtual environment using Red Hat Migration Analytics')}</EmptyState.Title>
    <EmptyState.Info>
      {__('Get a summarized report of your virtualization providers in CloudForms and collect a detailed inventory that can be further analyzed using the Migration Analytics application on cloud.redhat.com.') /* prettier-ignore */}
    </EmptyState.Info>
    <EmptyState.Action>
      <Button bsStyle="primary" bsSize="large" onClick={onGetStartedClick}>
        {__('Get started')}
      </Button>
    </EmptyState.Action>
  </EmptyState>
);

AnalyticsEmptyState.propTypes = {
  onGetStartedClick: PropTypes.func.isRequired
};

export default AnalyticsEmptyState;
