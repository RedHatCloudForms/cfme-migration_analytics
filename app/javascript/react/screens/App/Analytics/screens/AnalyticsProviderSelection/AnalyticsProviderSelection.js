import React from 'react';
import PropTypes from 'prop-types';
import { TypeAheadSelect, Button, Form, Icon } from 'patternfly-react';

const AnalyticsProviderSelection = ({
  summaryData,
  selectedProviders,
  selectProvidersAction,
  detailedDataSelected,
  selectDetailedDataAction,
  onContinueClick,
  onCancelClick
}) => (
  <Form>
    <p>{__('Select providers for which inventory data will be collected.')}</p>
    <TypeAheadSelect
      inputProps={{ id: 'providers-select' }}
      options={summaryData.providers}
      selected={selectedProviders}
      onChange={providers => selectProvidersAction(providers)}
      labelKey="name"
      placeholder={__('Select...')}
      emptyLabel={__('No matches found.')}
      multiple
      clearButton
      highlightOnlyResult
      selectHintOnEnter
    />
    <br />

    <Form.FormGroup className="detailed-data-selection">
      <Form.ControlLabel>{__('Inventory data needed')}</Form.ControlLabel>
      <Form.Radio
        checked={!detailedDataSelected}
        onChange={event => selectDetailedDataAction(event.target.value !== 'on')}
      >
        {__('Basic data')}
        <ul>
          <li>{__('High level information including number of VMs and hypervisors.')}</li>
          <li>{__('Necessary to calculate an initial migration savings estimate.')}</li>
          <li>{__('Data may or may not include SmartState Analysis results.')}</li>
        </ul>
      </Form.Radio>
      <Form.Radio
        checked={detailedDataSelected}
        onChange={event => selectDetailedDataAction(event.target.value === 'on')}
      >
        {__('Detailed data')}
        <ul>
          <li>
            {__('Detailed analysis of individual VMs including host, operating system, files, and running services.')}
          </li>
          <li>{__('Necessary for creating workload inventory and workload summary reports.')}</li>
          <li>
            {__('Requires that SmartState Analysis be run on each VM. This can take several hours depending on the number of VMs in the selected providers.') /* prettier-ignore */}
          </li>
        </ul>
        {detailedDataSelected && (
          <div className="icon-with-content smart-state-info">
            <Icon type="pf" name="warning-triangle-o" className="p-aligned" />
            <p>
              {__('See product documentation for instructions to configure and run SmartState Analysis.')}
              <br />
              <a href="/miq_task">{__('Progress of Smart State Analysis tasks can be monitored on the Tasks page.')}</a>
            </p>
          </div>
        )}
      </Form.Radio>
    </Form.FormGroup>

    <div className="footer-buttons">
      <Button bsStyle="primary" onClick={onContinueClick} disabled={selectedProviders.length === 0}>
        {__('Continue')}
      </Button>
      <Button onClick={onCancelClick}>{__('Cancel')}</Button>
    </div>
  </Form>
);

const providerShape = PropTypes.shape({
  id: PropTypes.number,
  name: PropTypes.string
});

AnalyticsProviderSelection.propTypes = {
  summaryData: PropTypes.shape({
    providers: PropTypes.arrayOf(providerShape)
  }).isRequired,
  selectedProviders: PropTypes.arrayOf(providerShape).isRequired,
  selectProvidersAction: PropTypes.func.isRequired,
  detailedDataSelected: PropTypes.bool.isRequired,
  selectDetailedDataAction: PropTypes.func.isRequired,
  onContinueClick: PropTypes.func.isRequired,
  onCancelClick: PropTypes.func.isRequired
};

export default AnalyticsProviderSelection;
