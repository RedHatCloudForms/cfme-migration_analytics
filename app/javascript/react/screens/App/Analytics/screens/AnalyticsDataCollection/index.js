import { connect } from 'react-redux';
import AnalyticsDataCollection from './AnalyticsDataCollection';
import {
  startInventoryBundleAction,
  fetchBundleTaskAction,
  resetDataCollectionStateAction
} from '../../redux/analyticsActions';
import {
  selectIsPayloadReady,
  selectBundleError,
  selectBundleTaskHref,
  selectIsBundleTaskFinished,
  selectPayloadPath,
  selectPayloadHost,
  selectNumVms
} from '../../redux/analyticsSelectors';
import { getPayloadUrl } from './helpers';

const mapStateToProps = ({ migrationAnalytics: { analytics } }) => {
  const bundleError = selectBundleError(analytics);
  const isBundleTaskFinished = selectIsBundleTaskFinished(analytics);
  const payloadPath = selectPayloadPath(analytics);
  const payloadHost = selectPayloadHost(analytics);
  return {
    selectedProviders: analytics.selectedProviders,
    bundleError,
    bundleTaskHref: selectBundleTaskHref(analytics),
    isFetchingBundleTask: analytics.isFetchingBundleTask,
    isBundleTaskFinished,
    isPayloadReady: selectIsPayloadReady(analytics, { bundleError, isBundleTaskFinished }),
    numVms: selectNumVms(analytics),
    payloadHost,
    payloadPath,
    payloadUrl: getPayloadUrl(payloadPath)
  };
};

export default connect(
  mapStateToProps,
  { startInventoryBundleAction, fetchBundleTaskAction, resetDataCollectionStateAction }
)(AnalyticsDataCollection);
