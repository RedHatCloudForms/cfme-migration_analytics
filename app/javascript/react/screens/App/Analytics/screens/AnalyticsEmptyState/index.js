import AnalyticsEmptyState from './AnalyticsEmptyState';

export default AnalyticsEmptyState;
