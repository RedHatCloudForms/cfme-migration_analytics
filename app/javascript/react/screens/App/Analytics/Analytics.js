import React from 'react';
import PropTypes from 'prop-types';
import { Breadcrumb, noop } from 'patternfly-react';
import Toolbar from './components/Toolbar';
import AnalyticsEmptyState from './screens/AnalyticsEmptyState';
import AnalyticsSummary from './screens/AnalyticsSummary';
import AnalyticsProviderSelection from './screens/AnalyticsProviderSelection';
import AnalyticsDataCollection from './screens/AnalyticsDataCollection';
import ManifestVersion from './components/ManifestVersion';
// import BreadcrumbPageSwitcher from '../common/BreadcrumbPageSwitcher'; // TODO: figure out how to share the breadcrumb switcher with v2v

const SCREENS = {
  EMPTY_STATE: 'EMPTY_STATE',
  SUMMARY: 'SUMMARY',
  PROVIDER_SELECTION: 'PROVIDER_SELECTION',
  DATA_COLLECTION: 'DATA_COLLECTION'
};

class Analytics extends React.Component {
  state = { currentScreen: SCREENS.EMPTY_STATE };

  componentDidMount() {
    this.props.fetchManifestInfoAction();
  }

  goToSummary = () => this.setState({ currentScreen: SCREENS.SUMMARY });
  goToProviderSelection = () => this.setState({ currentScreen: SCREENS.PROVIDER_SELECTION });
  goToDataCollection = () => this.setState({ currentScreen: SCREENS.DATA_COLLECTION });

  renderCurrentScreen = () => {
    const screens = {
      [SCREENS.EMPTY_STATE]: <AnalyticsEmptyState onGetStartedClick={this.goToSummary} />,
      [SCREENS.SUMMARY]: <AnalyticsSummary onCollectInventoryClick={this.goToProviderSelection} />,
      [SCREENS.PROVIDER_SELECTION]: (
        <AnalyticsProviderSelection onContinueClick={this.goToDataCollection} onCancelClick={this.goToSummary} />
      ),
      [SCREENS.DATA_COLLECTION]: (
        <AnalyticsDataCollection onCancelClick={this.goToProviderSelection} onReturnClick={this.goToSummary} />
      )
    };
    return screens[this.state.currentScreen];
  };

  render() {
    const { manifestInfo, toggleManifestUpdateModalAction, manifestUpdateModalVisible } = this.props;
    const { currentScreen } = this.state;
    const onEmptyState = currentScreen === SCREENS.EMPTY_STATE;
    return (
      <React.Fragment>
        <Toolbar>
          <Breadcrumb.Item active>{__('Compute')}</Breadcrumb.Item>
          <Breadcrumb.Item active>{__('Migration')}</Breadcrumb.Item>
          <Breadcrumb.Item active>
            <strong>{__('Migration Analytics')}</strong>
          </Breadcrumb.Item>
          {/* <BreadcrumbPageSwitcher activeHref="#/analytics" /> */}
          {/* TODO: figure out how to share the breadcrumb switcher with v2v */}
        </Toolbar>
        <div id="migration-analytics" className={onEmptyState ? 'row cards-pf' : ''}>
          {this.renderCurrentScreen()}
        </div>
        <ManifestVersion
          manifestInfo={manifestInfo}
          onEmptyState={onEmptyState}
          toggleManifestUpdateModalAction={toggleManifestUpdateModalAction}
          manifestUpdateModalVisible={manifestUpdateModalVisible}
          updatingManifest={false} // TODO
        />
      </React.Fragment>
    );
  }
}

Analytics.propTypes = {
  fetchManifestInfoAction: PropTypes.func,
  manifestInfo: PropTypes.object,
  toggleManifestUpdateModalAction: PropTypes.func,
  manifestUpdateModalVisible: PropTypes.bool
};

Analytics.defaultProps = {
  fetchManifestInfoAction: noop,
  manifestInfo: null,
  toggleManifestUpdateModalAction: noop,
  manifestUpdateModalVisible: false
};

export default Analytics;
