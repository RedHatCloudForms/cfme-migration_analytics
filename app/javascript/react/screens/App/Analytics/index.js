import { connect } from 'react-redux';
import Analytics from './Analytics';
import { fetchManifestInfoAction } from './redux/analyticsActions';

const mapStateToProps = ({
  migrationAnalytics: {
    analytics: { manifestInfo }
  }
}) => ({ manifestInfo });

export default connect(
  mapStateToProps,
  { fetchManifestInfoAction }
)(Analytics);
