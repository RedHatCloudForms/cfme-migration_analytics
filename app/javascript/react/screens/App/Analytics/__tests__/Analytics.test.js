import React from 'react';
import { shallow } from 'enzyme';
import Analytics from '../Analytics';
import AnalyticsEmptyState from '../screens/AnalyticsEmptyState';
import AnalyticsSummary from '../screens/AnalyticsSummary';
import AnalyticsProviderSelection from '../screens/AnalyticsProviderSelection';
import AnalyticsDataCollection from '../screens/AnalyticsDataCollection';

describe('Analytics root component', () => {
  const shallowSummaryScreen = () => {
    const component = shallow(<Analytics />);
    component
      .find(AnalyticsEmptyState)
      .props()
      .onGetStartedClick();
    return component;
  };

  const shallowProviderSelectionScreen = () => {
    const component = shallowSummaryScreen();
    component
      .find(AnalyticsSummary)
      .props()
      .onCollectInventoryClick();
    return component;
  };

  const shallowDataCollectionScreen = () => {
    const component = shallowProviderSelectionScreen();
    component
      .find(AnalyticsProviderSelection)
      .props()
      .onContinueClick();
    return component;
  };

  test('renders toolbar and empty state correctly', () => {
    const component = shallow(<Analytics />);
    expect(component.find(AnalyticsEmptyState)).toHaveLength(1);
    expect(component).toMatchSnapshot();
  });

  test('renders container with cards-pf class for empty state', () => {
    const component = shallow(<Analytics />);
    expect(component.find('AnalyticsContainer').dive()).toMatchSnapshot();
  });

  test('renders container without cards-pf class for other screens', () => {
    const component = shallowSummaryScreen();
    expect(component.find('AnalyticsContainer').dive()).toMatchSnapshot();
  });

  test('renders summary after clicking Get Started from empty state', () => {
    const component = shallowSummaryScreen();
    expect(component.find(AnalyticsSummary)).toHaveLength(1);
  });

  test('renders provider selection after clicking Collect Inventory from summary', () => {
    const component = shallowProviderSelectionScreen();
    expect(component.find(AnalyticsProviderSelection)).toHaveLength(1);
  });

  test('renders summary after clicking Cancel from provider selection', () => {
    const component = shallowProviderSelectionScreen();
    component
      .find('Connect(AnalyticsProviderSelection)')
      .props()
      .onCancelClick();
    expect(component.find(AnalyticsSummary)).toHaveLength(1);
  });

  test('renders data collection after clicking Continue on provider selection', () => {
    const component = shallowDataCollectionScreen();
    expect(component.find(AnalyticsDataCollection)).toHaveLength(1);
  });

  test('renders provider selection after clicking Cancel on data collection', () => {
    const component = shallowDataCollectionScreen();
    component
      .find('Connect(AnalyticsDataCollection)')
      .props()
      .onCancelClick();
    expect(component.find(AnalyticsProviderSelection)).toHaveLength(1);
  });

  test('renders summary after clicking Return to Summary on data collection', () => {
    const component = shallowDataCollectionScreen();
    component
      .find('Connect(AnalyticsDataCollection)')
      .props()
      .onReturnClick();
    expect(component.find(AnalyticsSummary)).toHaveLength(1);
  });
});
