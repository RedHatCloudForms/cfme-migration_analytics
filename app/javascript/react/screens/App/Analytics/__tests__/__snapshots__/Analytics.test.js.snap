// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Analytics root component renders container without cards-pf class when not on empty state screen 1`] = `
<div
  className=""
  id="migration-analytics"
>
  <Connect(AnalyticsSummary)
    onCollectInventoryClick={[Function]}
  />
</div>
`;

exports[`Analytics root component renders toolbar and empty state correctly 1`] = `
<Fragment>
  <Toolbar>
    <BreadcrumbItem
      active={true}
    >
      Compute
    </BreadcrumbItem>
    <BreadcrumbItem
      active={true}
    >
      Migration
    </BreadcrumbItem>
    <BreadcrumbItem
      active={true}
    >
      <strong>
        Migration Analytics
      </strong>
    </BreadcrumbItem>
  </Toolbar>
  <div
    className="row cards-pf"
    id="migration-analytics"
  >
    <AnalyticsEmptyState
      onGetStartedClick={[Function]}
    />
  </div>
  <ManifestVersion
    className="on-empty-state"
    isLoadingManifest={false}
    manifestError={null}
    manifestInfo={null}
    manifestUpdateModalVisible={false}
    resetManifestAction={[Function]}
    toggleManifestUpdateModalAction={[Function]}
    uploadManifestAction={[Function]}
  />
</Fragment>
`;
